src/Microsoft.ML.Data/Transforms/ValueToKeyMappingTransformer.cs

```diff
@@ -261,12 +261,12 @@ private static (string input, string output)[] GetColumnPairs(ColumnInfo[] colum
             return columns.Select(x => (x.Input, x.Output)).ToArray();
         }
 
-        internal string TestIsKnownDataKind(ColumnType type)
+        private string TestIsKnownDataKind(ColumnType type)
         {
             VectorType vectorType = type as VectorType;
             ColumnType itemType = vectorType?.ItemType ?? type;
 
-            if (itemType.RawKind != default && (vectorType != null || type is PrimitiveType))
+            if (itemType.RawType.IsValidDataKindType() && (vectorType != null || type is PrimitiveType))
```

Comment:
> && (vectorType != null || type is PrimitiveType) [](start = 54, length = 49)

I could be wrong, but it seems like many of these `IsValidDataKindType` are actually trying to capture the logic of what you've already written in your `IsStandardScalar` extension method on `ColumnType`. That is, "does it have one of the known datakinds," which is functionally the same as "is it one of the known builtin types."

(Also, incidentally I do not believe, if this first test succeeds, that it is possible for the statement after the `&&` to be false.)

Reply:
> many of these IsValidDataKindType are actually trying to capture the logic of what you've already written in your IsStandardScalar extension method on ColumnType

That's a good call. I forgot about that method. I've removed `IsValidDataKindType` and am using `IsStandardScalar` instead.

>I do not believe, if this first test succeeds, that it is not possible for the statement after the && to be false.

Agreed. Removed the 2nd expression.

---
In reply to: [248936996](https://github.com/dotnet/machinelearning/pull/2176#discussion_r248936996) [](ancestors = 248936996)

Reply:
Actually this broke a test. There is a test that passes in a `Vec<Key<U4>>` into this method. Thus, `RawType.IsValidDataKindType` and `IsStandardScalar` are not equivalent because the first allows for KeyTypes.

---
In reply to: [249152590](https://github.com/dotnet/machinelearning/pull/2176#discussion_r249152590) [](ancestors = 249152590,248936996)

Reply:
Huh interesting. Anyway I like your solution to the problem I see below.

---
In reply to: [249206559](https://github.com/dotnet/machinelearning/pull/2176#discussion_r249206559) [](ancestors = 249206559,249152590,248936996)